subsys/pm/power.c

```diff
@@ -273,12 +273,17 @@ enum pm_state pm_system_suspend(int32_t ticks)
 	return z_power_state.state;
 }
 
-void pm_notifier_register(struct pm_notifier *notifier)
+int pm_notifier_register(struct pm_notifier *notifier)
 {
+	if (notifier == NULL) {
+		return -EINVAL;
+	}
```

Comment:
Function doesn't document that it allows `NULL` (invalid pointer) as an input, so the caller can do this check.

Reply:
It doesn't deny  either. A NULL notifier will cause crash at line 184(sys_slist_apend). Actually get a NULL pointer is caller's fault, the caller should check it, but use a pointer without checking is this function's fault.
In this function there are four ways to handle a NULL parameter:
1. do nothing, system crash
2. add assert, system panic
3. return without return value
4. return with return value
Fail to register a pm notifier is not a big deal, it worth nothing to crash or panic at this point, way 1, 2 are the worst choice. As a common practice, fuctions like xxxxx_register() which are usually used in initialization process should return a value to help the process to make decision. On the other hand, functions like xxxx_unregister() (e. g. pm_notifier_unregister()) should not return any value, as they are used in release process which might has not enough resource to handle return values.

Reply:
> Fail to register a pm notifier is not a big deal, it worth nothing to crash or panic at this point, way 1, 2 are the worst choice

Calling `pm_notifier_register` with `NULL` is _undefined behavior_ if one reads the function docs, so it's up to the application to make sure valid data is passed. I'd say (1) and (2) are the most common way of _handling_ such problem in tree, not sure why "they are the worst choice".

Reply:
Thank you for your reply. I read pm.h and source code to understand  pm_notifier_register(), where can I find the function docs?